### ISSUE TITLE
Stubs created by shallow mounts do not show props

### ISSUE THREAD

**Issue by yakobe** [vuejs/vue-test-utils-next#213]
When using shallow mount, the stubs do not have the props anymore. In VTU for vue 2 a snapshot would show the props passed to a stubbed component. Now they are not there anymore and the snapshot diff's show them as removed.

Perhaps this is a configuration problem, or maybe this is not implemented yet? Do you have any info for us?

**Comment by lmiller1990** [comment]
We are passing this props to stubs [see here](https://github.com/vuejs/vue-test-utils-next/blob/31106332602e8afdc2c02d844743f167b2f3ffc7/src/stubs.ts#L127)

I am not exactly sure of the specific differences between how props are represented in Vue 2 and Vue 3 in the VDOM (and by extension VTU v1 and VTU v2 when serialized).

If this is only a problem for snapshots, I would just suggest updating your snapshots. Relying on props in snapshots (especially with shallowMount) to catch regressions is probably not a good practice.

**Comment by gem007bd** [comment]
@TheJaredWilcurt do you have any solution for this issue?

**Comment by TheJaredWilcurt** [comment]
Hey, I have not had time to get a working proof-of-concept up of combining Vue 3 + VTU 2 for my serializer library. So unfortunately I am not familiar enough with the changes in these libraries to be of help currently. However, when `jest-serializer-vue-tjw` is compatible with Vue 3/VTU 2, I do plan to add as many features as possible (similar to the Vue 2/VTU1 version).

If you can help me get past the barrier of getting Vue 3 and VTU to work, that would unlock a lot of work:

* https://github.com/tjw-lint/jest-serializer-vue-tjw/pull/64

**Comment by yakobe** [comment]
Hi @lmiller1990 
Thanks for the reply. Just for my understanding is https://github.com/tjw-lint/jest-serializer-vue-tjw the serializer that the snapshots are using by default? Or is it a replacement one that we could use to get the print out the props in vue3?

**Comment by lmiller1990** [comment]
That is an alternatively serializer for Vue 2. I am not sure what is in involved with making one for Vue 3 (not a snapshot user at all). You could try it and see what happens.

**Comment by yakobe** [comment]
Good to know. So the way the snapshots are serialized has nothing to do with the VTU's? Do you know if it is something controlled in vue3?

Sorry for bugging, just trying to work out where the problem lies to try and find out where to focus ☺️. We were using the snapshots to ensure props were being passed correctly. If they are not there then I'm sure others will have a similar issue. If it's down to vue3 itself then maybe I can find others there to collaborate with.

**Comment by lmiller1990** [comment]
You could accomplish the same thing with a regular Vue 3 + Jest + Snapshot setup, so I think it's not VTU specific. My understanding is you pass is something to Jest with the `toMatchSnapshot` matcher (eg `expect(whatever).toMatchSnapshot)`. the serializer will do whatever you tell it to with the passed-in value (in this case, `whatever`, which will normally be `wrapper.html()`.

As you noticed, props like `:foo="someComputedProperty"` will not be shown in the `html()` - makes sense, you don't see those in the inspector in your browser either. It only shows static props in the `html()`. 

To get things like computed props in the snapshot, you will likely need to do `expect(wrapper.vm).toMatchSnapshot`. In your serializer, you then do whatever to `wrapper.vm`, and the output is your snapshot. I have no idea if there is some way to dynamically access all the `props` on `wrapper.vm` to make the kind of snapshot you are describing.

If you need something special in VTU to make this process easierr, this is something can definitely look at adding something, but VTU doesn't have any special access to the `vm` (and things like `props`, `computed` etc).

**Comment by yakobe** [comment]
Thanks for the great explanation. I am now understand a bit more where the problems are 👍. Unfortunately my core development skills are more backend orientated 😄, but i'm totally willing to help out here where i can.

What i don't really understand is why there is a change to the props on the stubs. For example:

This is what the template in the `shallowMount` tested component:

```vue
<LabelUser :username="user.username" />
```

in vue2 we would get something like this in the snapshop
```html
<labeluser-stub username=\\"Dave\\"></labeluser-stub>
```

in vue3 we get:
```html
<label-user-stub></label-user-stub>
```

It's not the best or most secure test in the world, but with the props we can see that the value is passed in. 

I also notice that the name of the stub changed, although that not a big deal.

**Comment by lmiller1990** [comment]
I am not sure... it looks like we are passing the props to the stubs: https://github.com/vuejs/vue-test-utils-next/blob/c836ed00932e299c747c685380d3642ffdfc0a34/src/stubs.ts#L162 (this file is a bit complicated).

I will need to learn a little more about Jest serializers before advising any further.

Reading the stub name - this was probably an unintentional change. What do you think? I like the new one better, it matches the Vue convention of `MyComp` or `my-comp` both as valid component names.

**Comment by yakobe** [comment]
Sorry for the delay.

Yes, that file is a bit complicated. It's weird that it seems the props are passed on but the serializer doesn't print it. Also strange: if i create a stub manually using the `global.stubs` then it seems to print them out.

Also, if i add extra props that dont exist on a component and shallowMount it they are also printed. It seems that shallowMount just does not print out the props that actually exist 🤷 

The prop name change s just a minor inconvenience. We can update all the snapshots if we need to. Ideally the snapshots would need as little changing from vue2 to vue3 for the least amount of migration strain, but whatever is possible 😄

**Comment by lmiller1990** [comment]
Do you have other tests other than snapshots? Are those enough to give you confidence when upgrading?

Different people have different opinions on snapshot serialization. If you write a snapshot serializer and need something in VTU to make it work, we can work on something.

While snapshots can be useful (highlighting changes when you change your own code) I don't know how reliable they will be for something like updating from Vue 2 -> Vue 3, where not only the core lib but all the surrounding libs have more or less be rewritten.

**Comment by yakobe** [comment]
Yes, we have lots of other test. There tends to be 1 or 2 "render" tests each component that `shallowMount` and checks a snapshot looks ok, props are passed to child components etc. Then there are other tests that check for specific functionality. The problem is that a lot of the props etc were checked just in snapshots. 

Therefore, if this is removed then a large portion of our tests are gone. We didn't see an easier way to ensure that the child got the right props, since there can be lots of child components. I would be interested to know how others solved this sort of test requirement...?

**Comment by lmiller1990** [comment]
I think this needs to be implemented as a Jest Snapshot Serializer - Vue Test Utils is test runner agnostic, but snapshots are Jest specific. This might be useful information. I originally thought we should recommend using `html` for the snapshot - but this clearly has limitations, like showing props that may (or may not) appear in the `html`, like `:value:` from `v-model` etc.

Are you doing `expect(wrapper.html()).toMatchSnapshot()` or `expect(wrapper).toMatchSnapshot()`? Just using `html` means we are at the mercy of `vm.$el.html`; if we match on the `wrapper` itself, a serializer could access `vm` and do whatever needs to happen to show whatever props need to be shown. 

Either way, if this is valuable to some people, it *can* be done, but this is not something Vue Test Utils can do - I am fairly sure this needs to happen in a Jest serializer, since that is what a snapshot uses to create the snapshot to match against. I am pretty busy right now so I can't look into it immediately, if you or someone else reading this wants to explore this I can help out, though. It looks like there [is some work happening here](https://github.com/tjw-lint/jest-serializer-vue-tjw/pull/64).

______

My opinion on snapshots (and why I don't use them) is as follows.

Just knowing the child components got props is not enough, at least for me. I like to do a full render (with `mount`) and assert against whatever the child component is using the props for. For example in this `<Parent>`:

```html
<template>
  <Child msg="hello" />
</template>
```

Assuming `Child` is like this:

```html
<template>
  <span>Message is {{ msg }}</span>
</template>
```

I might do:

```js
const wrapper = mount(Parent)
expect(wrapper.html()).toContain('Message is hello')
```

If you deleted the `{{ msg }}` from `<Child>`, though, and were just using a snapshot on `shallowMoun(Parent)`, the test would continue to pass - despite the `{{ msg }}` not getting rendered.

Basically, knowing a function (or component) got the right props is not enough, in my experience - you need to assert on the returned value, in this case the DOM, to be 100% confident things are working as expected from a user point of view.

**Comment by yakobe** [comment]
I understand you opinion on the mount vs. shallowMount and find it interesting to see how different people approach different problems.

Our premise for the tests is that each "child" component usually has it's own test so we dont nee to test it in the parent and are happy knowing it just got the right data. Otherwise, if the child changes slightly we may end up having to update hundreds of tests where it is used. If we only check it gets the right data, we only have to update if the data it gets changes.

Thanks for the info so far. I will check out some of the serializer stuff.

**Comment by lmiller1990** [comment]
Cool, I will try to make some time to explore Jest serializers soon too, it's an interesting topic. I will post my progress and learning here. happy to work with you to support this feature if it's valuable to you.

**Comment by TheJaredWilcurt** [comment]
If someone can get a functioning Vue 3 + VTU 2 serializer plugin working, let me know, I'll be able to use that to get Jest-Serializer-Vue-TJW up and running for Vue 3.

Existing stalled attempt:

* tjw-lint/jest-serializer-vue-tjw#64

**Comment by lmiller1990** [comment]
Oh, I figured this out, I got stubs to render props. I will make a PR soon! I had to do some hacks in here: https://github.com/vuejs/vue-test-utils-next/blob/c836ed00932e299c747c685380d3642ffdfc0a34/src/stubs.ts#L162

```
  console.log
    <my-label-stub val="username"></my-label-stub>
```

I also wrote a super basic serializer, basically just like this:

```
    expect.addSnapshotSerializer({
      test(wrapper: VueWrapper<any>) {
        return '__app' in wrapper
      },
      serialize(wrapper: VueWrapper<any>) {
        return wrapper.html()
      }
    })
```

Either way, the problem is in that `stubs.ts` file, we are not passing the [props here](https://github.com/vuejs/vue-test-utils-next/blob/c836ed00932e299c747c685380d3642ffdfc0a34/src/stubs.ts#L33).

**Comment by yakobe** [comment]
@lmiller1990 You superstar! Thank you 😊

**Comment by lmiller1990** [comment]
I will release this one soon. We should figure out what we are doing in terms of snapshot serializers. Either we make the one by @TheJaredWilcurt the "official" one (in this case we need to upgrade to Vue 3 asap) or we do a discussion on what the ideal snapshot looks like and work towards that.
